specs/span-batches.md

```diff
@@ -0,0 +1,206 @@
+  - timestamp (~4 bytes)
+- The organization of block data is inefficient:
+  - Similar attributes are far apart, diminishing any chances of effective compression.
+  - Random data like hashes are positioned in-between the more compressible application data.
+- The RLP encoding of the data adds unnecessary overhead
+  - The outer list does not have to be length encoded, the attributes are known
+  - Fixed-length attributes do not need any encoding
+  - The batch-format is static and can be optimized further
+- Remaining meta-data for consistency checks can be optimized further:
+  - The metadata only needs to be secure for consistency checks. E.g. 20 bytes of a hash may be enough.
+
+Span-batches address these inefficiencies, with a new batch format version.
+
+## Span batch format
+
+Note that span-batches, unlike previous V0 batches,
+encode *a range of consecutive* L2 blocks at the same time.
+
+Introduce version `1` to the [batch-format](./derivation.md#batch-format) table:
+
+| `batch_version` | `content`           |
+|-----------------|---------------------|
+| 1               | `prefix ++ payload` |
+
+Notation:
+`++`: concatenation of byte-strings.
+`anchor`: first L2 block in the span
+`uvarint`: unsigned Base128 varint, as defined in [protobuf varint spec]
+
+[protobuf spec]: https://protobuf.dev/programming-guides/encoding/#varints
+
+Where:
+
+- `prefix = rel_timestamp ++ parent_check + l1_origin_check`
+  - `rel_timestamp`: relative time since genesis, i.e. `anchor.timestamp - config.genesis.timestamp`.
+  - `parent_check`: first 20 bytes of parent hash, i.e. `anchor.parent_hash[:20]`.
+  - `l1_origin_check`: to ensure the intended L1 origins of this span of
+        L2 blocks are consistent with the L1 chain, the blockhash of the last L1 origin is referenced.
+        The hash is truncated to 20 bytes for efficiency, i.e. `anchor.l1_origin.hash[:20]`.
+- `payload = block_count ++ block_tx_counts ++ tx_data_headers ++ tx_data + tx_sigs`:
+  - `block_count`: `uvarint` number of L2 blocks.
+  - `origin_bits`: bitlist of `block_count` bits, right-padded to a multiple of 8 bits:
+    1 bit per L2 block, indicating if the L1 origin changed this L2 block.
+  - `block_tx_counts`: for each block, a `uvarint` of `len(block.transactions)`.
+  - `tx_data_headers`: lengths of each `tx_data` entry, encodes as concatenated `uvarint` entries, (empty if there are
+    no entries).
+  - `tx_data`: [EIP-2718] encoded transactions.
+    - The `tx_signature` is truncated from each [EIP-2718] encoded tx. To be reconstructed from `tx_sigs`.
+      - `legacy`: starting at `v` RLP field
+      - `1` ([EIP-2930]): starting at `signatureYParity` RLP field
+      - `2` ([EIP-1559]): starting at `signature_y_parity` RLP field
+  - `tx_sigs`: concatenated list of fixed-length transactions:
+    - `legacy` txs `v` value is reduced to `y_parity` (chain ID is already known)
+    - `v` is encoded as `uint8`
```

Comment:
Legacy transactions aren't required to encode the chain ID in v. So I think we ether need the full `v` or both `y_parity` and a second boolean for whether the chain ID was specified or not.

But I'm not understanding why `v` is encoded as `uint8` if it's being reduced to a bool...

Reply:
Removed the assumption around chain-id, now just encoding the full `v` value.